timemachine/cpp/src/kernels/k_nonbonded_pairs.cuh

```diff
@@ -0,0 +1,208 @@
+#pragma once
+
+// NOTE: The order of associative operations performed by this kernel
+// should be identical to that of k_nonbonded. This is to ensure that
+// we get exact cancellation when subtracting exclusions computed
+// using this kernel.
+
+#include "../fixed_point.hpp"
+#include "k_fixed_point.cuh"
+#include "nonbonded_common.cuh"
+
+template <typename RealType>
+void __global__ k_nonbonded_pairs(
+    const int M, // number of pairs
+    const double *__restrict__ coords,
+    const double *__restrict__ params,
+    const double *__restrict__ box,
+    const double *__restrict__ dp_dl,
+    const double *__restrict__ coords_w, // 4D coords
+    const double *__restrict__ dw_dl,    // 4D derivatives
+    const int *__restrict__ pair_idxs,   // [M, 2] pair-list of atoms
+    const double *__restrict__ scales,   // [M]
+    const double beta,
+    const double cutoff,
+    unsigned long long *__restrict__ du_dx,
+    unsigned long long *__restrict__ du_dp,
+    unsigned long long *__restrict__ du_dl_buffer,
+    unsigned long long *__restrict__ u_buffer) {
+
+    // (ytz): oddly enough the order of atom_i and atom_j
+    // seem to not matter. I think this is because distance calculations
+    // are bitwise identical in both dij(i, j) and dij(j, i) . However we
+    // do need the calculation done for exclusions to perfectly mirror
+    // that of the nonbonded kernel itself. Remember that floating points
+    // commute but are not associative.
+
+    const int pair_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (pair_idx >= M) {
+        return;
+    }
+
+    int atom_i_idx = pair_idxs[pair_idx * 2 + 0];
+
+    RealType ci_x = coords[atom_i_idx * 3 + 0];
+    RealType ci_y = coords[atom_i_idx * 3 + 1];
+    RealType ci_z = coords[atom_i_idx * 3 + 2];
+    RealType ci_w = coords_w[atom_i_idx];
+
+    RealType dq_dl_i = dp_dl[atom_i_idx * 3 + 0];
+    RealType dsig_dl_i = dp_dl[atom_i_idx * 3 + 1];
+    RealType deps_dl_i = dp_dl[atom_i_idx * 3 + 2];
+    RealType dw_dl_i = dw_dl[atom_i_idx];
+
+    unsigned long long gi_x = 0;
+    unsigned long long gi_y = 0;
+    unsigned long long gi_z = 0;
+
+    int charge_param_idx_i = atom_i_idx * 3 + 0;
+    int lj_param_idx_sig_i = atom_i_idx * 3 + 1;
+    int lj_param_idx_eps_i = atom_i_idx * 3 + 2;
+
+    RealType qi = params[charge_param_idx_i];
+    RealType sig_i = params[lj_param_idx_sig_i];
+    RealType eps_i = params[lj_param_idx_eps_i];
+
+    unsigned long long g_qi = 0;
+    unsigned long long g_sigi = 0;
+    unsigned long long g_epsi = 0;
```

Comment:
I've eyeballed for local self-consistency (e.g. that L42-68 is identical to L70-96, and that the lines to zip (`x`, `+ 0`), (`y`, `+ 1`), (`z`, `+ 2`) etc. seem right) and this all looks look fine!

Not required, but I find long stretches of manually unrolled index arithmetic tedious to verify (and sometimes they contain subtle typos that sneak past review). Would be a small quality-of-life improvement if we can find a more concise idiom for `.x`, `.y`, `.z` that doesn't introduce runtime cost...

Reply:
This makes sense to me. To be clear, the implementation here was copied mostly unmodified from `k_nonbonded.cuh`, so I don't think there would be any new bugs introduced in the index arithmetic. But I'm definitely in favor of finding a zero (or acceptable) cost abstraction to make correctness more evident.